pallets/transfer-allowlist/src/lib.rs

```diff
@@ -0,0 +1,481 @@
+	}
+
+	pub type DepositBalanceOf<T> = <<T as Config>::ReserveCurrency as Currency<
+		<T as frame_system::Config>::AccountId,
+	>>::Balance;
+
+	//
+	// Storage
+	//
+	pub type AllowanceDetailsOf<T> = AllowanceDetails<<T as frame_system::Config>::BlockNumber>;
+
+	/// Struct to define when a transfer should be allowed from
+	/// the sender, receiver, and currency combination.
+	/// Transfer allowed time set by range of block numbers
+	/// Defaults to `allowed_at` starting at 0, and `blocked_at` ending at MAX block value
+	/// as per `Default` impl.
+	/// Current block must be between allowed at and blocked at
+	/// for transfer to be approved if allowance for sender/currency/receiver present.
+	#[derive(Clone, Debug, Encode, Decode, Eq, PartialEq, Default, MaxEncodedLen, TypeInfo)]
+	pub struct AllowanceDetails<BlockNumber> {
+		/// Specifies a block number after which transfers will be allowed
+		/// for the sender & currency and destination location.
+		/// This is by default set to 0 with the add allowance extrinsic,
+		/// unless a delay is set, in which case it is set to the current block + delay.
+		pub allowed_at: BlockNumber,
+		/// Specifies a block number after-which transfers will be blocked
+		/// for the sender & currency and destination location.
+		/// This is by default set to `BlockNumber::Max()`, except when an allowance has been removed but not purged.
+		/// In that case it is set to the current block + delay.
+		/// if the allowance is later updated with the add allowance extrinsic, it is set back to max.
+		pub blocked_at: BlockNumber,
+	}
+
+	impl<BlockNumber> AllowanceDetails<BlockNumber>
+	where
+		BlockNumber: AtLeast32BitUnsigned,
+	{
+		fn default() -> Self {
+			Self {
+				allowed_at: BlockNumber::zero(),
+				blocked_at: BlockNumber::max_value(),
+			}
+		}
+	}
+
+	/// Storage item for whether a sending account and currency have restrictions set
+	/// a double map is used here as we need to know whether there is a restriction set
+	/// for the account and currency.
+	/// Using an StorageNMap would not allow us to look up whether there was a restriction for the sending account and currency, given that:
+	/// - we're checking whether there's an allowance specified for the receiver location
+	///   - we would only find whether a restriction was set for the account in this caseif:
+	///     - an allowance was specified for the receiving location, which would render blocked restrictions useless
+	/// - we would otherwise need to store a vec of locations, which is problematic given that there isn't a set limit on receivers
+	/// If a transfer restriction is in place, then a second lookup is done on
+	/// AccountCurrencyAllowances to see if there is an allowance for the reciever
+	/// This allows us to keep storage map vals to known/bounded sizes.
+	#[pallet::storage]
+	#[pallet::getter(fn sender_currency_restriction_set)]
+	pub type AccountCurrencyTransferRestriction<T: Config> = StorageDoubleMap<
```

Comment:
Wouldnt be possible to merge this storage with `AccountCurrencyDelay` and have a tuple value (u64, BlockNumber). There should be an storage reduction based on the storage prefix+key.

Reply:
👍 Merged storage items